cmd/libsnap-confine-private/privs.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "privs.h"
+
+#define _GNU_SOURCE
+
+#include <unistd.h>
+
+#include <grp.h>
+#include <stdbool.h>
+#include <sys/capability.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "utils.h"
+
+static bool sc_has_capability(const char *cap_name)
+{
+	// Lookup capability with the given name.
+	cap_value_t cap;
+	if (cap_from_name(cap_name, &cap) < 0) {
+		die("cannot resolve capability name %s", cap_name);
+	}
+	// Get the capability state of the current process.
+	cap_t caps;
+	if ((caps = cap_get_proc()) == NULL) {
+		die("cannot obtain capability state (cap_get_proc)");
+	}
+	// Read the effective value of the flag we're dealing with
+	cap_flag_value_t cap_flags_value;
+	if (cap_get_flag(caps, cap, CAP_EFFECTIVE, &cap_flags_value) < 0) {
+		die("cannot obtain value of capability flag (cap_get_flag)");
+	}
+	// Free the representation of the capability state of the current process.
+	if (cap_free(caps) < 0) {
+		die("cannot free capability flag (cap_free)");
+	}
+	// Check if the effective bit of the capability is set.
+	return cap_flags_value == CAP_SET;
+}
+
+void sc_privs_drop()
+{
+	// Get the real, effective and saved user identifiers
+	uid_t ruid, euid, suid;
+	if (getresuid(&ruid, &euid, &suid) < 0) {
+		die("cannot get real, effective and saved user identifiers");
+	}
+	// Ditto for group identifiers
+	gid_t rgid, egid, sgid;
+	if (getresgid(&rgid, &egid, &sgid) < 0) {
+		die("cannot get real, effective and saved group identifiers");
+	}
+	if (euid == 0) {
```

Comment:
As mentioned, please get rid of this conditional. If non-root, the setgid() and setuid() calls are harmless. Consider if euid != 0 but egid == 0 for some reason. We still want to drop.